pkg/kubectl/resource/mapper.go

```diff
@@ -54,19 +56,21 @@ func (m *Mapper) InfoForData(data []byte, source string) (*Info, error) {
 	if err != nil {
 		return nil, fmt.Errorf("unable to decode %q: %v", source, err)
 	}
-	var obj runtime.Object
-	var versioned runtime.Object
-	if isThirdParty, gvkOut, err := thirdpartyresourcedata.IsThirdPartyObject(data, gvk); err != nil {
-		return nil, err
-	} else if isThirdParty {
-		obj, err = runtime.Decode(thirdpartyresourcedata.NewDecoder(nil, gvkOut.Kind), data)
-		versioned = obj
-		gvk = gvkOut
-	} else {
-		obj, versioned = versions.Last(), versions.First()
-	}
-	if err != nil {
-		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
+
+	obj, versioned := versions.Last(), versions.First()
+	if !m.dynamic {
```

Comment:
I'm not sure yet whether you have to fix this for your PR, but a dynamic resource without versioning and a third party resource are functionally equivalent.  I think the interfaces we have today should be almost sufficient to hide all of this from resource builder, which is why I bring it up.  I would have expect this pr to use helper (or identify what needs to be changed), provide the appropriate codec and clients down, and hide all of these details (preferably removing this code from knowing about third party).

We already have a way in codec to signal that objects aren't versioned, if versioned.First == versioned.Last.  I think that's all you need here to collapse dynamic objects with 3rd party.  And DynamicObjects() should handle third party resources (no generic command needs to know the difference).